### PR TITLE
Add `long` option to config file (equivalent to --long)

### PR DESCRIPTION
**Pull request by jfmontanaro**
I ran into the same issue as #621, expecting that explicitly-specified `blocks` in the config file would result in them being displayed by default. IMO it would be nice to have a config option for this - I prefer the config file over aliasing `ls` to `lsd -l` in my shell, because that way all of my settings are collected in one place.

And having it right below the `blocks` section in the example makes it harder to misunderstand how `blocks` works, the way I and the other commenter did.

Happy to make any changes, or just drop it if you don't feel this would be a good add. `alias ls='lsd -l'` _does_ work, after all.

---
#### TODO

- [x] Use `cargo fmt`
- [x] Add necessary tests
- [x] Add changelog entry
- [x] Update default config/theme in README (if applicable)
- [ ] ~~Update man page at lsd/doc/lsd.md (if applicable)~~

### CODE REVIEW

**Comment by meain** [comment]
The code changes looks good. Just one question. Do you always use `ls` with `-l` flag? From what I have personally seen, ~80% of `ls` use is without `-l` flag. This was one reason why `long` was never really made into a config option.

**Comment by jfmontanaro** [comment]
Weirdly enough yes, I do prefer `ls -l` in most situations. For some reason it feels more comfortable to me; I guess I just like to know the sizes and the mtime and such up front?

I think it may come from having spent a fair amount of time in either Powershell or [Nushell](https://nushell.sh), both of which default to a table-like display for their `ls` builtins. Arguably it makes a little more sense in those contexts because the output of the command is actually structured data rather than raw text, but I've gotten used to it just for informational use as well.

**Comment by zwpaper** [comment]
one more question, if long is set to true, do we have an args to change back to without `long` temporarily?

**Comment by jfmontanaro** [comment]
At present no, `long: true` behaves exactly like specifying `-l` on the command line. I can think of a few ways to address this:

* Add a dedicated option (e.g. `--short`) that cancels out `--long`. This seems undesirable since having `long: true` in the config file is such a niche use-case. I.e. for the vast majority of users an extra option like that would just be noise.
* Allow `-l/--long` to be _either_ a flag option _or_ an option that takes an explicit boolean value. That way you could to `--long false`, for instance, to disable it. However, I'm not sure if Clap permits options to function both ways like this - in fact I'm not even sure it's possible without major parsing ambiguity.
* Add a new option to specify the `layout` config item. Currently there's `--tree` and `--oneline`, which correspond to `Layout::Tree` and `Layout::OneLine` respectively, but there's no CLI option that explicitly corresponds to `Layout::Grid` (since it's the default, presumably). We could add `--grid` and make it override `--long`. This is a little more general than the above suggestion, and fits better into lsd's existing model, but it does still involve adding a new option.
* Co-opt an existing option, such as `--classic`. This has the downside that if implies a whole bunch more than just "don't display extended metadata", and it would be weird to force that baggage on the user when all they wanted was to disable the `--long` option.
* Or, alternatively: Just don't worry about the fact that there's no way to disable metadata display when you have `long: true` in the config file. Presumably, if you put it there, then you want it that way most of the time, and if you run into a situation where you don't then it's on you to use some hacky workaround like `/bin/ls` or `--config-file <(echo 'long: false')` or whatever.

**Comment by jfmontanaro** [comment]
Actually, the more I think about this the more I like the idea of allowing `--ignore-config` to take a value, so that you can specify _which_ config option to ignore. So in this case you could do `--ignore-config long`, and it would act as if the config file didn't have `long: true`. 

The nice thing about this is that it would be applicable to any config option (or at least, any top-level option). E.g. you could do `--ignore-config dereference` or something if you wanted to skip dereferencing for some reason.

`--ignore-config` by itself would still ignore the entire config file, of course.

**Comment by jfmontanaro** [comment]
Okay, scratch that. Turns out the only way to make an argument optionally take a value (i.e. `--ignore-config` and `--ignore-config=long` are both valid) is to set `arg.require_equals(true)`. This is at odds with how all the rest of the options are specified so I don't think it would be a good fit.

So now I'm back to thinking that the best option is a) to not worry about this problem and assume that people can use `--ignore-config` if they desperately need to disable `long: true`, or if that's not an option then b) to just drop this PR entirely as it's too hard to make mesh with the existing state of affairs.

**Comment by meain** [comment]
I'm still not sure weather we should add this or not. One options is to maybe rename the config option to `always-long` or `always-detailed`. This way it is more obvious that they can't change the turn off the option later. The behaviour remains the same, just with a different name.

**Comment by jfmontanaro** [comment]
That works for me, but I don't want to push too hard if you're uncomfortable with the add. Like I said this is a very minor convenience.

**Comment by zwpaper** [comment]
I am ok with adding the `always-long` option, I prefer `always-long` as most lsd(gnu ls) users should be familiar with the `long` option.

but like I said, we also need a command args to overwrite it, `ignore-config` seems to be too general, specifying something like `--grid` to overwrite `always-long` seem more reasonable

**Comment by eugenesvk** [comment]
You could override config with
 `--flag-` to disable and
`--flag+` to enable
And this could work for any flag, without the need to memorize  the `always` part, it's also much shorter

**Comment by zwpaper** [comment]
@eugenesvk 
> `--flag-` to disable and `--flag+` to enable

I don't think so, this is not *nix way, and most of our users are *nix users.

**Comment by jfmontanaro** [comment]
Ok, based on this discussion I have made the following changes:

* Rename `long` to `always-long` in config file
* Add `--grid` option to force non-long display even when set in config file

There are a couple aspects of this that might require further discussion so I'll highlight them here:

One, I've added an `ArgGroup` for `--grid`, `--tree` and `--oneline` as these options should be mutually exclusive. This seemed cleaner to me than adding individual `.conflicts_with()` calls to each arg. This doesn't seem to have any effect on the main help message, but it does affect error messages slightly. When you try to pass e.g. `--grid` and `--tree`:
```
error: The argument '--grid' cannot be used with '--tree'

USAGE:
    lsd.exe <--grid|--tree|--oneline>

For more information try --help
```
I don't think this is a problem, but I figured I should mention it since it affects user-facing behavior.

Two, the conditions in `flags::blocks::Blocks::configure_from` were getting a little hairy, so I reorganized the logic a bit. It now runs like this:
* Check for `--long` or `always-long: true`. If found, set blocks to the value found in the config file for `blocks`, or if not present in the config file use `Blocks::long()`. 
* Check for `--grid`. If found, set blocks to default, overriding the previous value.
* Check for explicitly-passed `--blocks` argument. If found, use that for `blocks`, again overriding the previous value.
* Check for `--context` and `--inode`. If found, prepend to `blocks`, whatever its value currently is.

All of the tests still pass (plus a few more that I added), so I'm relatively confident that this doesn't break anything. However, if someone with more experience of the codebase would like to check it over, that would probably be wise.

**Review by zwpaper — APPROVED** [review]
Hi @jfmontanaro, I have 2 more tiny questions, and I have no problems with this PR.

File: src/flags/blocks.rs [hunk 1]
```diff
@@ -153,10 +162,16 @@ impl Configurable<Self> for Blocks {
...
                 Some(Self(blocks))
             }
+        } else if let Some(true) = config.always_long {
+            Some(Self::long())
         } else {
             None
         }
     }
+
+    fn from_environment() -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (jfmontanaro) [reply src/flags/blocks.rs:172]: Whoops! Totally forgot I had added that. I think at one point I was experimenting with using the environment to override the config option rather than a CLI flag.

**Comment by meain** [comment]
`--grid` is something I wanted to have a different behavior. I like how in `exa` you can force grid mode but with more content. I was planning to hopefully mimic it in the future. This is a screenshot from how `exa` does it.



I'm not 100% certain about adding this in, but I don't think having `always-long` as a config option is that bad an idea. We can just leave it at that without any option to override it for now.

**Comment by jfmontanaro** [comment]
Ok, I've updated my fork to remove the `--grid` option. 

I've been doing some experimenting with using an env var to override `always_long`, e.g. setting `LSD_ALWAYS_LONG` set to `true` or `1` or `yes`. Code is [here](https://github.com/jfmontanaro/lsd/commit/d22ee488f9f93996dc21e93add208a80e002a977) if you want to take a look. 

I'm having trouble doing this cleanly, however. The signature of `Configurable::from_environment()` means that I can't use it to construct `Blocks`, because it would need to know the config file (so that it can decide what blocks to render) and that information isn't present in `from_environment()`. In _theory_ I could construct a `Blocks` with default values in `from_environment()`, then override those later in `configure_from()` based on the config file, but this seems like a potential footgun as most people would expect `from_environment()` to return a correct value for `Blocks` rather than a possibly-correct-but-might-need-to-be-overriden conditional value.

Also, because `Layout` changes depending on whether or not metadata is displayed, allowing that to be customized via the environment introduces some duplication of logic between `Layout` and `Blocks`. This was always the case, the env var just increases the level of duplication.

TL;DR: Using an env var to override `always_long` is possible, but messy.

**Comment by jfmontanaro** [comment]
Seems like this doesn't fit well into the existing structure so I'll go ahead and close it. Thanks for a great tool regardless!
